src/textual/widgets/_option_list.py

```diff
@@ -0,0 +1,904 @@
+        separator = lines_from(Rule(style=""))[0]
+
+        # Work through each item that makes up the content of the list,
+        # break out the individual lines that will be used to draw it, and
+        # also set up the tracking of the actual options.
+        line = 0
+        option = 0
+        for content in self._contents:
+            if isinstance(content, Option):
+                # The content is an option, so render out the prompt and
+                # work out the lines needed to show it.
+                new_lines = [
+                    Line(prompt_line, option)
+                    for prompt_line in lines_from(content.prompt, options)
+                ]
+                # Record the span information for the option.
+                add_span(OptionLineSpan(line, len(new_lines)))
+                if content.id is not None:
+                    # The option has an ID set, create a mapping from that
+                    # ID to the option so we can use it later.
+                    if content.id in option_ids:
+                        raise DuplicateID(
+                            f"The option list already has an option with id '{content.id}'"
+                        )
+                    option_ids[content.id] = option
+                option += 1
+            else:
+                # The content isn't an option, so it must be a separator (if
+                # there were to be other non-option content for an option
+                # list it's in this if/else where we'd process it).
+                new_lines = [Line(separator)]
+            add_lines(new_lines)
+            line += len(new_lines)
+
+        # Now that we know how many lines make up the whole content of the
+        # list, set the virtual size.
+        self.virtual_size = Size(self.scrollable_content_region.width, len(self._lines))
+
+    def add_option(self, item: NewOptionListContent = None) -> Self:
+        """Add a new option to the end of the option list.
+
+        Args:
+            item: The new item to add.
+
+        Returns:
+            The `OptionList` instance.
+        """
+        # Turn any incoming value into valid content for the list.
+        content = self._make_content(item)
+        self._contents.append(content)
+        # If the content is a genuine option, add it to the list of options.
+        if isinstance(content, Option):
+            self._options.append(content)
+        self._refresh_content_tracking(force=True)
+        self.refresh()
+        return self
+
+    def _remove_option(self, index: int) -> None:
+        """Remove an option from the option list.
+
+        Args:
+            index: The index of the item to remove.
+
+        Raises:
+            IndexError: If there is no option of the given index.
+        """
+        option = self._options[index]
+        del self._options[index]
+        del self._contents[self._contents.index(option)]
+        self._refresh_content_tracking(force=True)
```

Comment:
Similar question here: couldn't we "just" go over the options that show up after this one and offset their starting positions?

Reply:
Same answer here.